libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -169,10 +169,13 @@ proc getOrCreatePeer*(
 proc handleData*(p: PubSub, topic: string, data: seq[byte]): Future[void] {.async.} =
   if topic notin p.topics: return # Not subscribed
 
-  for h in p.topics[topic].handler:
+  # gather all futures without yielding to scheduler
+  let futures = p.topics[topic].handler.mapIt(it(topic, data))
+  # await each future and check for errors
+  for fut in futures:
```

Comment:
`allFinished()` - we can inspect each future for errors after, because otherwise we're essentially forcing sequential execution.

Reply:
ok, wonder what happens on cancelled tho...

Reply:
> otherwise we're essentially forcing sequential execution.

not quite, the execution remains concurrent even with the for loop - `allFinished` ensures that all work has completed before iterating over the finished futures so this ends up being _less_ concurrent (because the work in the for loop cannot start "early") - arguably, it's sometimes better to be less concurrent because there  are fewer opportunities for bugs.

also, when one future is cancelled, it's likely _all_ remaining futures should be cancelled, though this is tricky - @cheatfate might have some better insights here

Reply:
Well, yeah we spawn the futures concurrently in the `for` above and we're only awaiting sequentially so I guess it's a fair correction.